README.md

```diff
@@ -1,486 +1,679 @@
+`@bound` could be used with a mixin superclass as follows:
 
 ```js
-class C {
-  @wrap(f) method() { }
+class C extends WithBoundMethod(Object) {
+  #x = 1;
+  @bound method() { return this.#x; }
 }
+
+let c = new C;
+let m = c.method;
+m();  // 1, not TypeError
 ```
 
-is roughly equivalent to the following:
+This decorator could be defined as:
 
 ```js
-class C {
-  method() { }
+const boundName = Symbol("boundName");
+function bound(method, context) {
+  context.metadata = {[boundName]: true};
+  return method;
+}
+let boundMap = new MetadataLookupCache(boundName);
+
+function WithBoundMethods(superclass) {
+  return class C extends superclass {
+    constructor(...args) {
+      super(...args);
+      let names = boundMap.get(new.target, C);
+      for (const {name} of names) {
+        this[name] = this[name].bind(this);
+      }
+    }
+  }
 }
-C.prototype.method = f(C.prototype.method);
 ```
 
-`@wrap` can also be used on a class to wrap the entire class.
+Note the common use of `MetadataLookupCache` across both examples; this proposal or a follow-on one should consider adding a standard library for accing metadata for this purpose.
 
-```js
-@wrap(f)
-class C { }
-```
+### Option B: The `init` contextual keyword for methods
 
-is roughly equivalent to:
+If it's not acceptable to require a superclass/mixin for cases requiring initialization action, an The `init` keyword in a method declaration changes a method into an "init method". This keyword allows decorators to add initialization actions, run when the constructor executes.
 
-```js
-class C { }
-C = f(C);
-```
+#### `@on` with `init`
 
-Details:
-- `@wrap` may be used on private methods as well as public ones, static as well as instance.
-- The function is only passed the method, and no other context.
-- The return value is used to replace the method or accessor.
-- `@wrap` may be used on getters or setters, and applies to these individually.
-- `@wrap` may not be used on field declarations, as there's no clear meaning.
-- When `@wrap` is used on a class, if there is a use of `C` in a method or field initializer inside the class, it will refer to the original, unwrapped `C. See [#211](https://github.com/tc39/proposal-decorators/issues/211) for details.
+Usage:
 
-### `@register`
+```js
+class MyElement extends HTMLElement {
+  @on('click') init clickHandler() { }
+}
+```
 
-The `@register` decorator schedules a callback to run after the class is created.
+An "init method" (method declared with `init`) is called similarly to a method decorator, but it is expected to return a pair `{method, initialize}`, where `initialize` is called with the `this` value being the new instance, taking no arguments and returning nothing. 
 
 ```js
-class C {
-  @register(f) method() { }
+function on(eventName) {
+  return (method, context) => {
+    assert(context.kind === "init-method");
+    return {method, initialize() { this.addEventListener(eventName, method); }};
```

Comment:
What if instead of a new keyword the decorator implementation itself decided this method needed an initializer? If decorators were to always return an object, they could add optional fields like `metadata` or `initialize()` to trigger these behaviors. At least in the context of this specific example I don't see any merit in letting the user decide if the initializer should be added.

Reply:
The reason for not allowing all decorators to add this callback is because it would complicate the logic in constructors in the common case (where there is no initializer added), which is likely to cause runtime overhead in interpreters.

Reply:
Whether it is iterating through an array, or (as you and I were apparently both thinking) a conditional per decorated method in the desugaring for the constructor, runtime overhead is incurred in the interpreter per instance if a class has decorated methods when most methods don't need to use this functionality.

Reply:
Well, with this implementation it would only iterate through methods that return `initialize()`, skipping the rest. But surely you'd have to iterate through these in the constructor. I mean, how else would you initialize init methods if you don't?

Reply:
Oh, so the idea is to store each initialize function in a separate variable and then call it unconditionally in the constructor? I see now.